constraint/pkg/client/client_test.go

```diff
@@ -1039,6 +1041,261 @@ func TestClient_AddConstraint(t *testing.T) {
+						"foo": {
+							Type:    "array",
+							Default: &defaultsJSON,
+							Items: &apiextensions.JSONSchemaPropsOrArray{
+								Schema: &apiextensions.JSONSchemaProps{
+									Type: "string",
+								},
+							},
+						},
+						"bar": {
+							Type:    "array",
+							Default: &defaults2JSON,
+							Items: &apiextensions.JSONSchemaPropsOrArray{
+								Schema: &apiextensions.JSONSchemaProps{
+									Type: "string",
+								},
+							},
+						},
+					},
+				},
+			},
+			wantSpec: map[string]interface{}{
+				"parameters": map[string]interface{}{
+					"foo": defaults,
+					"bar": defaults2,
+				},
+			},
+			driver: regoDriver,
+		},
+		// this is the most interesting test as it shows:
+		// - two parameters defaulting
+		// - one parameter that is defined in the constraint AND does not default
+		{
+			name: "defaults for two, one mixed",
+			validationSpec: &templates.Validation{
+				OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+					Type: "object",
+					Properties: map[string]apiextensions.JSONSchemaProps{
+						"foo": {
+							Type:    "array",
+							Default: &defaultsJSON,
+							Items: &apiextensions.JSONSchemaPropsOrArray{
+								Schema: &apiextensions.JSONSchemaProps{
+									Type: "string",
+								},
+							},
+						},
+						"bar": {
+							Type:    "array",
+							Default: &defaults2JSON,
+							Items: &apiextensions.JSONSchemaPropsOrArray{
+								Schema: &apiextensions.JSONSchemaProps{
+									Type: "string",
+								},
+							},
+						},
+						"fuzz": {
+							Type:    "string",
+							Default: &defaultBuzzJSON,
+						},
+					},
+				},
+			},
+			// note that since the constraint is already creted when we modify it
+			// in place to simulate a constraint defining a fuzz parameter, we
+			// won't actually trigger a schema validation error.
+			constraintSpec: map[string]interface{}{
+				"parameters": map[string]interface{}{
+					"fuzz": "buzz",
+				},
+			},
+			wantSpec: map[string]interface{}{
+				"parameters": map[string]interface{}{
+					"foo":  defaults,
+					"bar":  defaults2,
+					"fuzz": "buzz",
+				},
+			},
+			driver: regoDriver,
+		},
+		{
+			name: "defaults; rego driver",
+			validationSpec: &templates.Validation{
+				OpenAPIV3Schema: &apiextensions.JSONSchemaProps{
+					Type: "object",
+					Properties: map[string]apiextensions.JSONSchemaProps{
+						"foo": {
+							Type:    "array",
+							Default: &defaultsJSON,
+							Items: &apiextensions.JSONSchemaPropsOrArray{
+								Schema: &apiextensions.JSONSchemaProps{
+									Type: "string",
+								},
+							},
+						},
+					},
+				},
+			},
+			wantSpec: map[string]interface{}{
+				"parameters": map[string]interface{}{
+					"foo": defaults,
+				},
+			},
+			driver: regoDriver,
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			tc := tc
+			// given a template and constraint for it
+			template := template.DeepCopy()
```

Comment:
let's not rely on shadowing. Let's be explicit, call the right side something like "baseTemplate", same for "constraint" below.